### PR TITLE
move all variables in sampleAndWaterMarkHistograms::innerSet

### LINKED ISSUE #97685: performance regression on ARM: 1.20 apiserver has high cpu usage
**What happened**:
Cluster was upgraded from 1.19.5 to 1.20.1 using kubeadm. After that, kube-apiserver cpu usage went from 50% to 300%.

**What you expected to happen**:
Expected that apiserver has the same cpu usage.

**How to reproduce it (as minimally and precisely as possible)**:
Update a cluster from 1.19.5 to 1.20.1.

**Anything else we need to know?**:
The cluster is running on sbc hardware, thats why the original 50% cpu usage, and thats why 300% cpu usage is noticeable and causes service disruption.

**Environment**:
- Kubernetes version (use `kubectl version`): (the apiserver is rolled back to 1.19.5)
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.5", GitCommit:"e338cf2c6d297aa603b50ad3a301f761b4173aa6", GitTreeState:"clean", BuildDate:"2020-12-09T11:10:32Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/arm"}
```
- Cloud provider or hardware configuration:
bare metal on SBC hardware.
- OS (e.g: `cat /etc/os-release`):
```
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

- Kernel (e.g. `uname -a`):
```
# uname -a
Linux k8s-master01 5.4.86 #1 SMP Wed Dec 30 17:58:36 UTC 2020 armv7l GNU/Linux
```
- Install tools:
kubeadm
- Network plugin and version (if this is a network-related bug):
kube-router
- Others:
Please check profiling of each apiserver.
[apiserver.tar.gz](https://github.com/kubernetes/kubernetes/files/5764262/apiserver.tar.gz)

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What type of PR is this?**

/kind bug
/kind regression

**What this PR does / why we need it**:
This is a refactoring to tiptoe around https://github.com/golang/go/issues/43570 for #97685

**Which issue(s) this PR fixes**:

Fixes #97685 

**Special notes for your reviewer**:
This is an alternative to #97791 and #97854

**Does this PR introduce a user-facing change?**:

```release-note
Fixes a performance regression in 1.19 on ARM architectures computing priority and fairness metrics
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by MikeSpreitzer** [comment]
@liggitt 
@erikwilson 
@brandond 
@tkashem 
@adtac

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go [hunk 1]
```diff
@@ -160,17 +160,14 @@ func (saw *sampleAndWaterMarkHistograms) SetX1(x1 float64) {
 }
 
 func (saw *sampleAndWaterMarkHistograms) innerSet(updateXOrX1 func()) {
-	var when time.Time
-	var whenInt int64
-	var acc sampleAndWaterMarkAccumulator
-	var wellOrdered bool
-	func() {
+	when, whenInt, acc, wellOrdered := func() (time.Time, int64, sampleAndWaterMarkAccumulator, bool) {
```
Comment (brandond) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:163]: At this point it seems like there's not much point in keeping it as an anonymous function, might as well just move that whole block out to `func (saw *sampleAndWaterMarkHistograms) validateQuantization(time.Time, int64, sampleAndWaterMarkAccumulator, bool)` (name it something better) unless the intent is to revert all of this once the golang bug is fixed.

**Review by brandond — CHANGES_REQUESTED** [review]

**Comment by MikeSpreitzer** [comment]
This fixes the problem for 386 on linux.

```
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git checkout test-saw
Switched to branch 'test-saw'
Your branch is up to date with 'origin/test-saw'.
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git branch -D test-and-fix
Deleted branch test-and-fix (was 91ef7229c36).
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git pull
remote: Enumerating objects: 33, done.
remote: Counting objects: 100% (33/33), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 33 (delta 27), reused 33 (delta 27), pack-reused 0
Unpacking objects: 100% (33/33), 3.00 KiB | 439.00 KiB/s, done.
From https://github.com/MikeSpreitzer/kubernetes
 + 714f40f3be9...14f94b6c5e5 rejigger-quantize-3  -> origin/rejigger-quantize-3  (forced update)
 * [new branch]              rejigger-quantize-3a -> origin/rejigger-quantize-3a
 * [new branch]              rejigger-quantize-4  -> origin/rejigger-quantize-4
Already up to date.
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git checkout -b test-and-fix
Switched to a new branch 'test-and-fix'
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git cherry-pick origin/rejigger-quantize-4
[test-and-fix d2613d7a69f] move all variables in sampleAndWaterMarkHistograms::innerSet
 Date: Fri Jan 8 13:32:38 2021 -0500
 1 file changed, 7 insertions(+), 9 deletions(-)
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ 
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ 
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ GOARCH=386 go test -c .
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ for i in {1..100}; do if GOGC=1 ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ for i in {1..100}; do if ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
```

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:163]: the point of the inner func is to unlock via a `defer` as soon as our calculations are done before we log or call out to record metrics

Comment (brandond) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:163]: The side effects argument makes sense I suppose. I get that it just exists to handle the unlock cleanly, but to me the code picks up a bit of an odor when anonymous functions grow return values and such. I'm just here as peanut gallery though, feel free to ignore my feedback.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:163]: I agree with @liggitt , the inner function should never be invoked in any other context and so I would prefer to not expose it.

**Comment by liggitt** [comment]
can `/hold cancel` once this is verified on ARM

**Comment by brandond** [comment]
I can verify on arm this afternoon, if nobody else beats me to it.

**Comment by liggitt** [comment]
@MikeSpreitzer if you want to go ahead and open the cherry-pick to release-1.20 so CI is happy, that would help meet the 1.20.2 deadline of EOD

**Review by hasheddan — APPROVED** [review]
/lgtm

@MikeSpreitzer is this ready to go?

**Comment by brandond** [comment]
Working for me on arm:

```
sysadm@pi02:kubernetes$ git log --pretty=oneline | head -n 3
ba7a0759472012798fc3862bcf1b63558f6df2c0 Explain warnings in new test
737d4e2cef29d1f2089452be36bf9d46423ef553 Add unit test for sample-and-watermark histograms
611184aa59d0cd40466bc3bc4b40a3712a038171 move all variables in sampleAndWaterMarkHistograms::innerSet
sysadm@pi02:kubernetes$ cd staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics
sysadm@pi02:metrics$ GOARCH=arm GOARM=7 go test -c .
sysadm@pi02:metrics$ for i in {1..100}; do if GOGC=1 ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
sysadm@pi02:metrics$ for i in {1..100}; do if ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
```

**Comment by liggitt** [comment]
thanks

/hold cancel

**Pull request merged**